integration/fuse/src/main/java/alluxio/fuse/file/FuseFileOutStream.java

```diff
@@ -0,0 +1,175 @@
+        fileLen = 0;
+        LOG.debug(String.format("Open path %s with flag 0x%x for overwriting. "
+            + "Alluxio deleted the old file and created a new file for writing", uri, flags));
+      } else {
+        // Support open(O_WRONLY flag) - truncate(0) - write() workflow, otherwise error out
+        return new FuseFileOutStream(fileSystem, authPolicy, Optional.empty(), fileLen, uri, mode);
+      }
+    }
+    return new FuseFileOutStream(fileSystem, authPolicy,
+        Optional.of(AlluxioFuseUtils.createFile(fileSystem, authPolicy, uri, mode)),
+        fileLen, uri, mode);
+  }
+
+  private FuseFileOutStream(FileSystem fileSystem, AuthPolicy authPolicy,
+      Optional<FileOutStream> outStream, long fileLen, AlluxioURI uri, long mode) {
+    Preconditions.checkNotNull(fileSystem);
+    Preconditions.checkNotNull(authPolicy);
+    Preconditions.checkNotNull(outStream);
+    Preconditions.checkNotNull(uri);
+    mFileSystem = fileSystem;
+    mAuthPolicy = authPolicy;
+    mOutStream = outStream;
+    mOriginalFileLen = fileLen;
+    mURI = uri;
+    mMode = mode;
+  }
+
+  @Override
+  public int read(ByteBuffer buf, long size, long offset) {
+    throw new UnsupportedOperationException("Cannot read from write only stream");
+  }
+
+  @Override
+  public synchronized void write(ByteBuffer buf, long size, long offset) {
+    if (size > Integer.MAX_VALUE) {
+      throw new UnsupportedOperationException(
+          String.format("Cannot write more than %s", Integer.MAX_VALUE));
+    }
+    if (!mOutStream.isPresent()) {
+      throw new UnsupportedOperationException(
+          "Cannot overwrite/extending existing file without O_TRUNC flag or truncate(0) operation");
+    }
+    int sz = (int) size;
+    long bytesWritten = mOutStream.get().getBytesWritten();
+    if (offset != bytesWritten && offset + sz > bytesWritten) {
+      throw new UnsupportedOperationException(String.format("Only sequential write is supported. "
+          + "Cannot write bytes of size %s to offset %s when %s bytes have written to path %s",
+          size, offset, bytesWritten, mURI));
+    }
+    if (offset + sz <= bytesWritten) {
+      LOG.warn("Skip writing to file {} offset={} size={} when {} bytes has written to file",
+          mURI, offset, sz, bytesWritten);
+      // To fulfill vim :wq
+    }
+    final byte[] dest = new byte[sz];
+    buf.get(dest, 0, sz);
+    try {
+      mOutStream.get().write(dest);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public synchronized long getFileLength() {
+    return mOutStream.map(FileOutStream::getBytesWritten).orElse(mOriginalFileLen);
+  }
+
+  @Override
+  public synchronized void flush() {
+    if (!mOutStream.isPresent()) {
+      return;
+    }
+    try {
+      mOutStream.get().flush();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public synchronized void truncate(long size) {
```

Comment:
just wondering if there is a concurrency problem with calls from different hosts? For example fuse1 is going to create or truncate one file, while fuse2 is also trying to operate the same file?

Reply:
Then it relies on Alluxio's mechanism to solve multi-user problems, but I think the correctness may not be able to be guaranteed in all scenarios. The same problem happens to all Alluxio APIs, but Fuse may be more severe because it contains the logic of truncate() and maybe others.
When two clients try to create the same file, Alluxio has its own mechanism to guarantee only one client wins. 
When one client tries to read, the other client tries to truncate and write, yeah, the read will likely error out but that's also expected similar to other alluxio APIs.
We can discuss the detailed read/write logic and see how to improve it. This one is more for refactoring purposes instead of fixing problems. Making the code cleaner so we can easier to understand the read/write logics and improve them.

Reply:
agree.